### DISCUSSION TITLE
Add arbitrary history state via `pushState` / `replaceState` API

### DISCUSSION THREAD

**Discussion by kcrwfrd** [vercel/next.js#58453] [Help]
Is there a way to add arbitrary data to [History API's state](https://developer.mozilla.org/en-US/docs/Web/API/History/state) (as done with `pushState`)?

I'm looking to replicate the back button functionality on Twitter. If you go to twitter.com home page, then click into a profile, then click into another profile, you can press the back button in Twitter's UI, and it will call `history.back()` as you can observe in your browser history.

If you open a new tab, go to google.com, and then navigate directly to a profile on twitter, like https://twitter.com/cameronmoll, and then you press the back button, it doesn't call `history.back()`, as that would take you to Google. Instead, it navigates to a fallback URL of `/home`.

Since you can't introspect the history stack to determine if the previous entry is on the same domain, I'm pretty sure they're achieving this by adding state via `pushState`. If you've navigated from another page on twitter and you check `history.state`, you'll see it has an object like this

```
{
    "key": "funbk9",
    "state": {
        "userId": "1250847387062419456",
        "fromApp": true,
        "previousPath": "/home"
    }
}
```

I wish to achieve the same functionality in our Next.js app

I could maybe hack something together with [sessionStorage](https://developer.mozilla.org/en-US/docs/Web/API/Window/sessionStorage)... unfortunately the challenge is I need to relate that state to the currently active history entry. I think I would have to listen to `popstate` events and try to determine if user has pressed the back or forward buttons to determine my position in the history stack... sounds pretty ugly and messy, whereas adding arbitrary state data via `pushState` is pretty nice and clean.

In this case I am using the pages router but am also wondering about with the app router, too.

**Comment by prkagrawal** [comment]
Weird I don't see this behaviour!


https://github.com/vercel/next.js/assets/49971882/327b5e98-3c59-457d-810a-0341769fadf4


Using firefox in arch

**Comment by kcrwfrd** [comment]
Sorry, I was referring to the back button within Twitter's UI. The button will either call `history.back()` if the app determines you navigated from another Twitter URL, or it will fall back to navigating to `/home`. Here's an example of the behavior:

https://github.com/vercel/next.js/assets/504158/942f42d0-0424-44d0-8b17-15a99fcbb5a1

**Comment by steve-marmalade** [comment]
I haven't looked into this functionality deeply yet, but did you see that some form of support for native History API that was added recently? https://nextjs.org/docs/app/building-your-application/routing/linking-and-navigating#using-the-native-history-api

**Reply by kcrwfrd** [reply]
Exciting, thank you for calling this out. I may be able to re-implement our own version of `'next/link'` that will add our own state via `pushState` / `replaceState` that would support the desired back button feature.

Another [decent suggestion](https://github.com/vercel/next.js/discussions/23991#discussioncomment-6552390) (however slightly unpalatable) is to monkey-patch `History.prototype.pushState` to include similar state information that I drew as an example from Twitter.

Of course, the upcoming [Navigation API](https://developer.mozilla.org/en-US/docs/Web/API/Navigation_API) will also solve the problem 😅

**Comment by Mossop** [comment]
As far as I can tell using the native history API doesn't work if you're also changing routes. The URL in the address bar changes correctly but the new route doesn't render.

**Reply by kcrwfrd** [reply]
@Mossop are you using the app router, or pages?

**Reply by Mossop** [reply]
Using the app router

**Reply by mikeKlech** [reply]
same
